### PR TITLE
[OCPCLOUD-1587] Validate discriminated unions via webhooks

### PR DESCRIPTION
**Pull request by JoelSpeed**
This PR is made up of 2 commits. The first of which is a prefactor to make the second commit easier to work with.

The new content here is a set of validations added so that discriminated unions are validated correctly. IE, when the discriminant is set, only the correct field within the union is set, no extras. This should prevent users from setting invalid configurations

### CODE REVIEW

**Review by damdo — COMMENTED** [review]
Nice refactor, just a non-blocking question on naming.
Additions around discriminated union validation look good.
/lgtm

File: pkg/webhooks/controlplanemachineset/webhooks.go [hunk 1]
```diff
@@ -363,3 +404,66 @@ func missingFailureDomains(list1 []failuredomain.FailureDomain, list2 []failured
...
+			errs = append(errs, field.Forbidden(parentPath.Child(fieldJSONName), fmt.Sprintf("value not allowed when %s is %q", discriminantJSONName, discriminantValue)))
+		}
+	}
+
+	return errs
+}
+
+// getJSONName gets the JSON name of the field from the struct tag.
+// This is used so that errors show the name as the end user would use.
+func getJSONName(typ reflect.StructField) string {
```
Comment (alexander-demichev) [inline pkg/webhooks/controlplanemachineset/webhooks.go:467]: what does `typ` mean?

Comment (JoelSpeed) [reply pkg/webhooks/controlplanemachineset/webhooks.go:467]: Yeah good point, I'm gonna rename this to `structField`

**Review by damdo — COMMENTED** [review]
/lgtm

**Review by alexander-demichev — APPROVED** [review]
/lgtm

**Comment by JoelSpeed** [comment]
/override ci/prow/e2e-aws-upgrade

This PR doesn't affect the E2Es yet as this isn't in payload

**Pull request merged**